### ISSUE TITLE
When watch cache can't establish a watch successfully for its backing store, all resourceVersion=0 queries start to pile up

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#49684]
On a very large cluster in etcd2 mode on 1.6 (100k secrets, ~500mb of data in JSON), LIST /api/v1/secrets takes about 40s (20s to fetch from etcd, 20s to decode and then encode back to v1 secret).  When the apiserver restarts, the combination of very long list and very short watch interval (100 writes/sec) means that the apiserver is never able to establish a watch.  This appears to cause the storage cacher to hold a write lock long enough that secret requests from the nodes pile up on the api server (trying to acquire a read lock from the storage cache) to fill max-inflight (set to 3k) until the lock is released, at which point the requests are served.  This continues indefinitely until the watch cacher gets lucky and is able to establish a watch.

Causes:
* Very large LIST calls (100k secrets @ 500mb that take > 40s to complete) 
* Large numbers of nodes with large number of unique secrets (250 nodes w/ 10k pods with 10-20k unique secrets @ 220 secret GET /sec from nodes)
* Max-in-flight that is larger than base load, but not larger than peak load (3k max in flight against 300-400 reads per sec)
* Etcd2, where watch window is consumed before list call completes (1k window, 100 writes/sec)

The watch cache in general is not resilient to this particular problem (cache reflectors thrash and do not back off when they can't establish lists), and whatever lock we are holding is too aggressive.

I'd like to implement paging #49338 for informers to mitigate this, but I think the reflector and watch cache also need to be resilient against the failure mode of being unable to establish a list watch.  etcd3 will help as well, but at 1m keys I think this may still be a risk.

@wojtek-t we are seeing this on large production clusters, I'm going to track down where in the watch cache we're getting hung up

**Comment by smarterclayton** [comment]
It appears that this is due to the ready condition (which holds all Watch, Get, List, and GetToList until the first sync completes).  I think we're right on the cusp of the 1m timeout for all API calls enforced by timeout handler - so ready never goes true.

I think the right thing to do is to punch through the cacher if it's not ready.  It adds load to etcd, but when the watch cache goes pathological it's the right thing to do.  A minor flourish would be to set a reasonable minimum time for the watch cache to be available (15s?) and then start punching through.

@liggitt @deads2k will post a PR for review tomorrow.  I was able to reproduce 100% by patching cacheListerWatcher in cacher.go:

```
// Implements cache.ListerWatcher interface.
func (lw *cacherListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
	glog.Infof("List start %s", lw.resourcePrefix)
	list := lw.newListFunc()
	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, "", Everything, list); err != nil {
		return nil, err
	}
	if lw.resourcePrefix == "/secrets" {
		time.Sleep(time.Duration(40+rand.Intn(15)) * time.Second)
	} else {
		time.Sleep(time.Duration(5+rand.Intn(10)) * time.Second)
	}
	glog.Infof("List end %s", lw.resourcePrefix)
	return list, nil
}

// Implements cache.ListerWatcher interface.
func (lw *cacherListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
	if lw.resourcePrefix == "/secrets" {
		if rand.Float32() < 0.9 {
			return nil, errors.NewGone("resource version has expired")
		}
	}
	return lw.storage.WatchList(context.TODO(), lw.resourcePrefix, options.ResourceVersion, Everything)
}
```

**Comment by wojtek-t** [comment]
> I think the right thing to do is to punch through the cacher if it's not ready.

Let me clarify this, because I'm not sure I 100% agree with it. 
If cacher is not ready, it's definitely the right decision to forward GET/LIST requests to etcd directly.

However, I don't think it's a good decision to do that for watches, because if we forward a number of watches to etcd, they we will end up deserializing a number of objects a number of times and we will starve apiserver and etcd even more.

I will send out a PR with that fix in few minutes.

**Comment by wojtek-t** [comment]
https://github.com/kubernetes/kubernetes/pull/49688 is out (it's not perfect, since we may probably want to wait for few seconds before doing this, but I don't have time to look into it more today).

**Comment by smarterclayton** [comment]
> However, I don't think it's a good decision to do that for watches, because if we forward a number of watches to etcd, they we will end up deserializing a number of objects a number of times and we will starve apiserver and etcd even more.

I have actually never seen this production system be CPU starved.  Instead, it is gated by all of our artificial limits :)  For this system we are running at 3 cores api server, 1 core controller manager+scheduler, and 2 cores etcd.  But we are completely and 100% rate bound by the size of the data set.  I think our SLO goals that led to the watch cache trying to avoid watches was too focused on high rate of change with low cardinality data sets - the cluster this is being observed on is exactly the opposite - low rate of change with high cardinality.  I'd say that we may want to potentially delay the watches, but I don't think blocking watches is 100% the right thing to do.  

Also, if the cache never initializes, the cluster is in limp mode.  In limp mode, the appropriate limit might be to reject watches or rate limit them, but I'm not terribly confident that in limp mode watches on etcd are the problem.  And I think if high rate of change in etcd (via a particular watch) is the signal, we might actually want to discriminate on who is making the call.

**Comment by wojtek-t** [comment]
> I have actually never seen this production system be CPU starved. 

We've seen :)

> Instead, it is gated by all of our artificial limits :) For this system we are running at 3 cores api server, 1 core controller manager+scheduler, and 2 cores etcd. But we are completely and 100% rate bound by the size of the data set. I think our SLO goals that led to the watch cache trying to avoid watches was too focused on high rate of change with low cardinality data sets - the cluster this is being observed on is exactly the opposite - low rate of change with high cardinality. I'd say that we may want to potentially delay the watches, but I don't think blocking watches is 100% the right thing to do.

> Also, if the cache never initializes, the cluster is in limp mode. In limp mode, the appropriate limit might be to reject watches or rate limit them, but I'm not terribly confident that in limp mode watches on etcd are the problem. And I think if high rate of change in etcd (via a particular watch) is the signal, we might actually want to discriminate on who is making the call.

I think we should weight it what is that watch and where it is coming from. If it is say 1000-node clusters and thos are watch Pods (and as you know etcd doesn't support filtering so any change of a Pod will be delivered to every watched and only filtered out in apiserver), I claim that it would kill us. If it is a single watch issued from some random client, then yes - I agree we should forward it to etcd.

Unfortunately for now I would say it's pretty difficult to distinguish those. So I'm a bit against of simply forwarding watchers to etcd (even delayed). But at the same time, I completely agree that the current state is far from ideal - I will try to think how we can improve it (but if you already have any suggestions, I'm more than happy to hear them).

**Comment by mml** [comment]
cc @jpbetz

**Comment by smarterclayton** [comment]
Fixed in #49688
